File: src/lib/errfile/describe.ts
// Turning a thrown value into text — pm/error_err.mdx §3.2.
//
// Everything here runs ONLY on the error path, and every function is total: a hostile or broken
// throwable (a throwing getter, a throwing toString, a circular cause chain) produces text, never
// an exception (R11). This file imports nothing: it is the bottom of the library.

/** The message is capped here (§3.2), cut in the middle so the start and the end both survive. */
export const MESSAGE_CAP = 2000;
/** How much of the start of an over-long message survives … */
export const MESSAGE_HEAD = 1000;
/** … and how much of the end. */
export const MESSAGE_TAIL = 990;
/** The data block is capped here. */
export const DATA_CAP = 1000;
/** The whole record (header + stack) is capped here. */
export const RECORD_CAP = 8000;
/** How far `err.cause` is walked. */
export const CAUSE_DEPTH = 5;
/** Stack frames kept after trimming. */
export const STACK_FRAMES = 12;

/** `Type: message (k=v …)` — one link of a cause chain, already turned into plain data. */
export type DescribedError = {
    type: string;
    message: string;
    codes: [string, string][];
};

// \x00-\x1f, \x7f, U+2028 and U+2029: anything that could end a line or forge a second header.
const CONTROL_CHARS = /[\u0000-\u001f\u007f\u2028\u2029]/g;

/** Replace every control character (and the two Unicode line separators) with a space. */
export function stripControlChars(value: string): string {
    return value.replace(CONTROL_CHARS, ' ');
}

/** The `⋯(N chars cut)⋯` marker put where a string was cut. */
function cutMarker(cut: number): string {
    return ` …(${cut} chars cut)… `;
}

/** Cap a message at MESSAGE_CAP: the first MESSAGE_HEAD and the last MESSAGE_TAIL characters survive. */
export function capMessage(value: string): string {
    if (value.length <= MESSAGE_CAP) {
        return value;
    }

    const cut = value.length - MESSAGE_HEAD - MESSAGE_TAIL;
    return value.slice(0, MESSAGE_HEAD) + cutMarker(cut) + value.slice(value.length - MESSAGE_TAIL);
}

/** Cap any string at `max` characters by cutting out its middle. */
export function capMiddle(value: string, max: number): string {
    if (value.length <= max) {
        return value;
    }

    const keep = Math.max(0, max - cutMarker(value.length).length);
    const head = Math.ceil(keep / 2);
    const tail = keep - head;
    const cut = value.length - keep;
    return value.slice(0, head) + cutMarker(cut) + (tail > 0 ? value.slice(value.length - tail) : '');
}

/** Read a property without letting a throwing getter escape. */
export function prop(value: unknown, key: string): unknown {
    if (value === null || (typeof value !== 'object' && typeof value !== 'function')) {
        return undefined;
    }

    try {
        return Reflect.get(value, key);
    } catch {
        return undefined;
    }
}

/** String(value), but total. */
export function safeString(value: unknown): string {
    try {
        if (typeof value === 'string') {
            return value;
        }

        if (value === undefined) {
            return 'undefined';
        }

        if (value === null) {
            return 'null';
        }

        if (typeof value === 'object') {
            try {
                const json = JSON.stringify(value);

                if (typeof json === 'string' && json !== '{}') {
                    return json;
                }
            } catch {
                // circular or throwing toJSON — fall through to String()
            }
        }

        return String(value);
    } catch {
        return '[unprintable value]';
    }
}

function isErrorLike(value: unknown): boolean {
    return value instanceof Error || (typeof value === 'object' && value !== null && typeof prop(value, 'message') === 'string');
}

/** The fields that become the `(code=… errno=…)` suffix, in order (§3.2). */
const CODE_KEYS = ['code', 'errno', 'op', 'syscall', 'HttpStatusCode', 'errorCode'];

function codesOf(err: unknown): [string, string][] {
    const codes: [string, string][] = [];

    if (typeof err !== 'object' || err === null) {
        return codes;
    }

    for (const key of CODE_KEYS) {
        const v = prop(err, key);

        if (typeof v === 'string' || typeof v === 'number') {
            codes.push([key, String(v)]);
        }
    }

    return codes;
}

/** One throwable → `{ type, message, codes }`, without its cause chain. Total. */
export function describeOne(err: unknown): DescribedError {
    try {
        if (isErrorLike(err)) {
            const rawName = prop(err, 'name');
            const rawMessage = prop(err, 'message');

            return {
                type: typeof rawName === 'string' && rawName ? rawName : 'Error',
                message: typeof rawMessage === 'string' ? rawMessage : '',
                codes: codesOf(err)
            };
        }

        return {
            type: typeof err === 'object' ? 'Thrown' : 'Thrown ' + typeof err,
            message: safeString(err),
            codes: []
        };
    } catch {
        return { type: 'Thrown', message: '[unprintable value]', codes: [] };
    }
}

/** The `.cause` chain of a throwable, outermost first, at most CAUSE_DEPTH links, cycle-safe. */
export function describeChain(err: unknown): DescribedError[] {
    const chain: DescribedError[] = [];
    const seen = new Set<unknown>([err]);
    let current = prop(err, 'cause');

    for (let depth = 0; depth < CAUSE_DEPTH && current !== undefined && current !== null; depth++) {
        if (seen.has(current)) {
            break;
        }

        seen.add(current);
        chain.push(describeOne(current));
        current = prop(current, 'cause');
    }

    return chain;
}

/** `Type: message (k=v …)` as text — the message stripped of control characters and capped. */
export function headlineText(d: DescribedError): string {
    const message = capMessage(stripControlChars(d.message));
    const head = message ? `${stripControlChars(d.type)}: ${message}` : stripControlChars(d.type);
    const codes = d.codes.map(([k, v]) => `${stripControlChars(k)}=${stripControlChars(v)}`);
    return codes.length ? `${head} (${codes.join(' ')})` : head;
}

/** ` | cause: …` for each link of a cause chain, at most CAUSE_DEPTH of them. */
export function causesText(chain: DescribedError[]): string {
    let out = '';

    for (const link of chain.slice(0, CAUSE_DEPTH)) {
        out += ` | cause: ${headlineText(link)}`;
    }

    return out;
}

/** Frames → indented `    at …` lines, at most STACK_FRAMES of them. */
export function stackText(frames: string[]): string {
    return frames.slice(0, STACK_FRAMES).map(f => `    at ${stripControlChars(f)}`).join('\n');
}

/**
 * The one-line description of a throwable: headline plus cause chain, no stack. For code that has
 * to SHOW a message — never to build a report by hand (R2). Secrets are not redacted here; use
 * `describeError` from index.ts for anything that leaves the process.
 */
export function describeThrowable(err: unknown): string {
    return headlineText(describeOne(err)) + causesText(describeChain(err));
}

// Absolute prefixes shortened to repo-relative: anything up to a `/src/`, `/mcp/src/` or
// `/mcp/dist/` segment, a file:// scheme, or a loopback origin (http://localhost:8081/).
const REPO_PREFIX = /(?:file:\/\/)?(?:\/[^\s():]*?)?\/(src\/|mcp\/(?:src|dist)\/)/g;
const ORIGIN_PREFIX = /\bhttps?:\/\/(?:localhost|127\.0\.0\.1|\[::1\]|0\.0\.0\.0)(?::\d+)?\//g;
// V8 `    at fn (file:line:col)` / `    at file:line:col`; Firefox/Safari `fn@file:line:col`.
const FRAME = /^\s*at\s|^[^\s@]*@\S+:\d+/;

function isOurOwnFrame(line: string): boolean {
    return line.includes('node_modules') ||
        line.includes('(node:internal/') ||
        line.includes(' node:internal/') ||
        line.includes('/errfile/') ||
        line.includes('/@vite/client');
}

/** Keep only frame lines, drop node_modules and our own frames, cap, shorten paths, indent. */
export function trimStack(stack: string, maxFrames: number = STACK_FRAMES): string {
    try {
        const frames: string[] = [];

        for (const raw of stack.split('\n')) {
            if (frames.length >= maxFrames) {
                break;
            }

            if (!FRAME.test(raw) || isOurOwnFrame(raw)) {
                continue;
            }

            const line = stripControlChars(raw.trim())
                .replace(ORIGIN_PREFIX, '')
                .replace(REPO_PREFIX, '$1');
            frames.push(`    ${line.startsWith('at ') ? line : 'at ' + line}`);
        }

        return frames.join('\n');
    } catch {
        return '';
    }
}

/** The trimmed stack of a throwable, or '' when it has none. */
export function stackOf(err: unknown): string {
    const stack = prop(err, 'stack');
    return typeof stack === 'string' ? trimStack(stack) : '';
}

File: pkg/errfile/describe.go
package errfile

import (
	"errors"
	"fmt"
	"net"
	"reflect"
	"runtime"
	"strconv"
	"strings"
	"syscall"
	"unicode/utf8"
)

// The caps of §3.2 (mirrored in testdata/vectors.json "caps").
const (
	// MessageCap is the longest message kept, cut in the middle.
	MessageCap = 2000
	// DataCap is the longest formatted data block.
	DataCap = 1000
	// RecordCap is the longest whole record (header plus stack).
	RecordCap = 8000
	// CauseDepth is how far the Unwrap chain is walked.
	CauseDepth = 5
	// JoinMembers is how many members of an errors.Join are shown.
	JoinMembers = 3
	// StackFrames is how many trimmed frames are kept.
	StackFrames = 12
)

// KV is one key=value pair of a code suffix or a data block, already redacted.
type KV struct {
	Key   string
	Value string
}

// stripControlChars replaces \x00-\x1f, \x7f, U+2028 and U+2029 with a space, so a field can never
// end a line or forge a second header (§3.2). Every other rune passes through unchanged.
func stripControlChars(s string) string {
	clean := true

	for _, r := range s {
		if isControl(r) {
			clean = false
			break
		}
	}

	if clean {
		return s
	}

	var b strings.Builder
	b.Grow(len(s))

	for _, r := range s {
		if isControl(r) {
			b.WriteByte(' ')
		} else {
			b.WriteRune(r)
		}
	}

	return b.String()
}

func isControl(r rune) bool {
	return r < 0x20 || r == 0x7f || r == 0x2028 || r == 0x2029
}

// capMiddle keeps a string within max characters by cutting out its middle, so the start and the
// end both survive. The head keeps max/2 characters, the tail max/2-10, and the marker says how
// many were cut (" …(416 chars cut)… "). Characters are runes, never split mid-sequence.
func capMiddle(s string, max int) string {
	if len(s) <= max {
		return s
	}

	n := utf8.RuneCountInString(s)

	if n <= max {
		return s
	}

	head := max / 2
	tail := max/2 - 10

	if tail < 0 {
		tail = 0
	}

	runes := []rune(s)
	cut := n - head - tail

	return string(runes[:head]) + " …(" + strconv.Itoa(cut) + " chars cut)… " + string(runes[n-tail:])
}

// Headline renders one link of an error as "Type: message (k=v …)", control-stripped and capped.
// The vectors test drives it directly; describeError builds the same text from a live error.
func Headline(typ, message string, codes []KV) string {
	text := stripControlChars(typ)

	if message != "" {
		text += ": " + capMiddle(stripControlChars(message), MessageCap)
	}

	return text + codeSuffix(codes)
}

func codeSuffix(codes []KV) string {
	if len(codes) == 0 {
		return ""
	}

	parts := make([]string, 0, len(codes))

	for _, kv := range codes {
		parts = append(parts, stripControlChars(kv.Key)+"="+stripControlChars(kv.Value))
	}

	return " (" + strings.Join(parts, " ") + ")"
}

// described is the text form of an error: its headline, its cause chain, and the parts the fold
// key and the transient-network check need.
type described struct {
	typ      string
	message  string
	headline string
	cause    string
	codes    []KV
}

// typeName is the dynamic type of an error, as %T prints it: *fs.PathError, syscall.Errno,
// *errs.Error, *errors.errorString.
func typeName(err error) string {
	return fmt.Sprintf("%T", err)
}

// safeMessage is err.Error(), but total: a panicking Error() becomes text, never a panic.
func safeMessage(err error) (msg string) {
	defer func() {
		if r := recover(); r != nil {
			msg = "[Error() panicked: " + fmt.Sprint(r) + "]"
		}
	}()

	if err == nil {
		return ""
	}

	if isNilPointer(err) {
		return "<nil>"
	}

	return err.Error()
}

func isNilPointer(v0 any) bool {
	v := reflect.ValueOf(v0)

	return v.Kind() == reflect.Pointer && v.IsNil()
}

// describeError walks the error: headline, then " | cause: …" per Unwrap link, up to CauseDepth.
// An errors.Join shows its first JoinMembers members as causes. Every string is redacted for URL
// query values and the statements root before it is returned.
func describeError(err error) described {
	d := described{}

	if err == nil {
		return d
	}

	d.typ = typeName(err)
	d.message = redactText(safeMessage(err))
	d.codes = codesOf(err)
	d.headline = Headline(d.typ, d.message, d.codes)

	seen := map[uintptr]bool{}
	remember(seen, err)
	depth := 0
	var b strings.Builder

	for _, link := range nextLinks(err) {
		walkCause(&b, link, seen, &depth)
	}

	d.cause = b.String()

	return d
}

func walkCause(b *strings.Builder, err error, seen map[uintptr]bool, depth *int) {
	if err == nil || *depth >= CauseDepth || isNilPointer(err) {
		return
	}

	if !remember(seen, err) {
		return // a cycle
	}

	*depth++
	b.WriteString(" | cause: ")
	b.WriteString(Headline(typeName(err), redactText(safeMessage(err)), codesOf(err)))

	for _, link := range nextLinks(err) {
		walkCause(b, link, seen, depth)
	}
}

// remember records a pointer-identified error in `seen`; it returns false when it was seen before.
// Non-pointer errors cannot form cycles, so they are always new.
func remember(seen map[uintptr]bool, err error) bool {
	v := reflect.ValueOf(err)

	if v.Kind() != reflect.Pointer || v.IsNil() {
		return true
	}

	p := v.Pointer()

	if seen[p] {
		return false
	}

	seen[p] = true

	return true
}

// nextLinks returns the causes one level down: Unwrap() error, or the first JoinMembers of
// Unwrap() []error.
func nextLinks(err error) (links []error) {
	defer func() {
		if r := recover(); r != nil {
			links = nil
		}
	}()

	switch u := err.(type) {
	case interface{ Unwrap() error }:
		if next := u.Unwrap(); next != nil {
			return []error{next}
		}
	case interface{ Unwrap() []error }:
		all := u.Unwrap()

		if len(all) > JoinMembers {
			all = all[:JoinMembers]
		}

		return all
	}

	return nil
}

// codesOf collects the code-like values of ONE link: a Code() method, exported fields named Code,
// Errno, Op, Syscall, HttpStatusCode or ErrorCode, and the symbolic name of a syscall.Errno.
func codesOf(err error) (codes []KV) {
	defer func() {
		if r := recover(); r != nil {
			codes = nil
		}
	}()

	if errno, ok := err.(syscall.Errno); ok {
		return []KV{{Key: "errno", Value: errnoName(errno)}}
	}

	if dns, ok := err.(*net.DNSError); ok {
		switch {
		case dns.IsNotFound:
			return []KV{{Key: "code", Value: "ENOTFOUND"}}
		case dns.IsTimeout:
			return []KV{{Key: "code", Value: "ETIMEDOUT"}}
		case dns.IsTemporary:
			return []KV{{Key: "code", Value: "EAI_AGAIN"}}
		}

		return nil
	}

	switch c := err.(type) {
	case interface{ Code() int32 }:
		codes = append(codes, KV{Key: "code", Value: strconv.FormatInt(int64(c.Code()), 10)})
	case interface{ Code() int }:
		codes = append(codes, KV{Key: "code", Value: strconv.Itoa(c.Code())})
	case interface{ Code() string }:
		if c.Code() != "" {
			codes = append(codes, KV{Key: "code", Value: c.Code()})
		}
	}

	v := reflect.ValueOf(err)

	for v.Kind() == reflect.Pointer || v.Kind() == reflect.Interface {
		if v.IsNil() {
			return codes
		}

		v = v.Elem()
	}

	if v.Kind() != reflect.Struct {
		return codes
	}

	t := v.Type()

	for i := 0; i < t.NumField(); i++ {
		f := t.Field(i)

		if !f.IsExported() {
			continue
		}

		key := ""

		switch strings.ToLower(f.Name) {
		case "code":
			key = "code"
		case "errno":
			key = "errno"
		case "op":
			key = "op"
		case "syscall":
			key = "syscall"
		case "httpstatuscode", "statuscode":
			key = "status"
		case "errorcode":
			key = "error_code"
		default:
			continue
		}

		if key == "code" && len(codes) > 0 && codes[0].Key == "code" {
			continue
		}

		fv := v.Field(i)
		val := ""

		switch fv.Kind() {
		case reflect.String:
			val = fv.String()
		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
			if fv.Int() != 0 {
				val = strconv.FormatInt(fv.Int(), 10)
			}
		case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
			if fv.Uint() != 0 {
				val = strconv.FormatUint(fv.Uint(), 10)
			}
		}

		if val != "" {
			codes = append(codes, KV{Key: key, Value: val})
		}
	}

	return codes
}

var errnoNames = map[syscall.Errno]string{
	syscall.EACCES:       "EACCES",
	syscall.EADDRINUSE:   "EADDRINUSE",
	syscall.EAGAIN:       "EAGAIN",
	syscall.ECONNABORTED: "ECONNABORTED",
	syscall.ECONNREFUSED: "ECONNREFUSED",
	syscall.ECONNRESET:   "ECONNRESET",
	syscall.EEXIST:       "EEXIST",
	syscall.EHOSTUNREACH: "EHOSTUNREACH",
	syscall.EINVAL:       "EINVAL",
	syscall.EIO:          "EIO",
	syscall.EISDIR:       "EISDIR",
	syscall.EMFILE:       "EMFILE",
	syscall.ENETUNREACH:  "ENETUNREACH",
	syscall.ENOENT:       "ENOENT",
	syscall.ENOSPC:       "ENOSPC",
	syscall.ENOTDIR:      "ENOTDIR",
	syscall.ENOTEMPTY:    "ENOTEMPTY",
	syscall.EPERM:        "EPERM",
	syscall.EPIPE:        "EPIPE",
	syscall.ETIMEDOUT:    "ETIMEDOUT",
}

func errnoName(e syscall.Errno) string {
	if name, ok := errnoNames[e]; ok {
		return name
	}

	return "errno" + strconv.Itoa(int(e))
}

// transientMarkers are the codes and message fragments of §11.2 (vectors "transient_network_markers").
var transientMarkers = []string{
	"ECONNREFUSED", "ENOTFOUND", "EAI_AGAIN", "ETIMEDOUT", "ECONNRESET", "ERR_NETWORK", "ECONNABORTED",
	"Network Error", "network unreachable",
}

var transientErrnos = map[syscall.Errno]bool{
	syscall.ECONNREFUSED: true, syscall.ECONNRESET: true, syscall.ETIMEDOUT: true,
	syscall.ECONNABORTED: true, syscall.EHOSTUNREACH: true, syscall.ENETUNREACH: true,
}

// isTransientNetwork reports whether an error (or anything it wraps, three links deep) is the
// "laptop lid closed" kind: a refused/reset/timed-out connection, a DNS miss, or a message that
// carries one of the markers. Such faults are WARN and fold per 10 minutes (§11.2).
func isTransientNetwork(err error) (transient bool) {
	defer func() {
		if r := recover(); r != nil {
			transient = false
		}
	}()

	current := err

	for depth := 0; depth < 3 && current != nil; depth++ {
		var errno syscall.Errno

		if errors.As(current, &errno) && transientErrnos[errno] {
			return true
		}

		var dns *net.DNSError

		if errors.As(current, &dns) && (dns.IsNotFound || dns.IsTemporary || dns.IsTimeout) {
			return true
		}

		var op *net.OpError

		if errors.As(current, &op) && op.Timeout() {
			return true
		}

		msg := safeMessage(current)

		for _, m := range transientMarkers {
			if strings.Contains(msg, m) {
				return true
			}
		}

		links := nextLinks(current)

		if len(links) == 0 {
			break
		}

		current = links[0]
	}

	return false
}

// ── stacks and source positions ────────────────────────────────────────────────────────────────

// ownFrames are function-name fragments whose frames are dropped from every stack and never chosen
// as `where` (§3.2): the runtime, this library (and its vendored copy), logrus and gin's dispatch.
var ownFrames = []string{
	"runtime.",
	"/pkg/errfile.",
	"/pkg/errfile/",
	"/cli/internal/errfile.",
	"/cli/internal/errfile/",
	"github.com/sirupsen/logrus.",
	"github.com/gin-gonic/gin.",
}

func isOwnFrame(function, file string, extra []string) bool {
	if strings.HasPrefix(function, "runtime.") {
		return true
	}

	// The library's own test files report like any caller.
	if !strings.HasSuffix(file, "_test.go") {
		for _, m := range ownFrames[1:] {
			if strings.Contains(function, m) {
				return true
			}
		}
	}

	for _, m := range extra {
		if m != "" && strings.HasPrefix(function, m) {
			return true
		}
	}

	return false
}

// frame is one captured stack frame after trimming.
type frame struct {
	function string // short: (*ingestApply).writePlan
	file     string // repo-relative
	line     int
}

func (f frame) String() string {
	return f.function + " (" + f.file + ":" + strconv.Itoa(f.line) + ")"
}

// captureFrames reads the goroutine's stack from `skip` frames above the caller and trims it:
// own frames (plus `extra` prefixes) are dropped; when `afterPanic` is set, everything up to and
// including runtime.gopanic is dropped too, so a panic site is the first frame. At most
// StackFrames frames are kept.
func captureFrames(skip int, extra []string, afterPanic bool) []frame {
	pcs := make([]uintptr, 64)
	n := runtime.Callers(skip+2, pcs)

	if n == 0 {
		return nil
	}

	iter := runtime.CallersFrames(pcs[:n])
	var out []frame
	dropping := afterPanic

	for {
		fr, more := iter.Next()

		if dropping {
			if strings.HasPrefix(fr.Function, "runtime.gopanic") || strings.HasPrefix(fr.Function, "runtime.panic") {
				dropping = false
			}

			if !more {
				break
			}

			continue
		}

		if fr.Function != "" && !isOwnFrame(fr.Function, fr.File, extra) {
			out = append(out, frame{function: shortFunction(fr.Function), file: relPath(fr.File), line: fr.Line})

			if len(out) >= StackFrames {
				break
			}
		}

		if !more {
			break
		}
	}

	return out
}

// shortFunction strips the import path and package name: "…/pkg/machine.(*ingestApply).writePlan"
// → "(*ingestApply).writePlan"; "main.main" → "main".
func shortFunction(fn string) string {
	if i := strings.LastIndex(fn, "/"); i >= 0 {
		fn = fn[i+1:]
	}

	if i := strings.Index(fn, "."); i >= 0 {
		fn = fn[i+1:]
	}

	return fn
}

// relPath makes a source path repo-relative (R14): the module path
// (github.com/mayswind/ezbookkeeping/ or …/ezbookkeeping/cli/ → cli/) and any absolute prefix up to
// and including /ezbookkeeping/ are stripped. Module-cache paths keep their module part.
func relPath(file string) string {
	file = strings.ReplaceAll(file, "\\", "/")

	if i := strings.LastIndex(file, "/ezbookkeeping/"); i >= 0 {
		return file[i+len("/ezbookkeeping/"):]
	}

	if strings.HasPrefix(file, "github.com/mayswind/ezbookkeeping/") {
		return strings.TrimPrefix(file, "github.com/mayswind/ezbookkeeping/")
	}

	if i := strings.Index(file, "/pkg/mod/"); i >= 0 {
		return file[i+len("/pkg/mod/"):]
	}

	if i := strings.Index(file, "/go/src/"); i >= 0 {
		return file[i+len("/go/src/"):]
	}

	for _, marker := range []string{"/pkg/", "/cmd/", "/cli/", "/scripts/"} {
		if i := strings.Index(file, marker); i >= 0 {
			return file[i+1:]
		}
	}

	return file
}

// whereFrom is the `where` of a report: the first kept frame as file.go:line.
func whereFrom(frames []frame) string {
	if len(frames) == 0 {
		return "?"
	}

	return frames[0].file + ":" + strconv.Itoa(frames[0].line)
}

// whereFile is `where` without its :line, for the fold key (R10: a loop reports from one line).
func whereFile(where string) string {
	if i := strings.LastIndex(where, ":"); i > 0 {
		if _, err := strconv.Atoi(where[i+1:]); err == nil {
			return where[:i]
		}
	}

	return where
}

// CallerWhere returns the repo-relative file:line of the nearest caller whose function is outside
// this library and outside the given function-name prefixes. The logrus hook uses it to point at
// the caller of log.Errorf rather than at pkg/log.
func CallerWhere(skipPrefixes ...string) string {
	return whereFrom(captureFrames(1, skipPrefixes, false))
}

// Describe renders an error as "Type: message (code=…) | cause: …" with no stack, for code that has
// to SHOW a message (a CLI line, an envelope). It must never be used to build a report by hand.
func Describe(err error) string {
	if err == nil {
		return ""
	}

	d := describeError(err)

	return d.headline + d.cause
}
